sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,695 @@
+  uint16_t rx_fifo_end = config->rx_fifo_len - 1;
+  uint16_t tx_fifo_start = config->rx_fifo_len;
+  uint16_t tx_fifo_end = config->rx_fifo_len + config->tx_fifo_len - 1;
+  if (tx_fifo_end >= kDifSpiDeviceSramLen) {
+    // We've overflown the SRAM region...
+    return kDifSpiDeviceResultBadArg;
+  }
+
+  uint32_t rx_fifo_bounds = 0;
+  rx_fifo_bounds |= (rx_fifo_start & SPI_DEVICE_RXF_ADDR_BASE_MASK)
+                    << SPI_DEVICE_RXF_ADDR_BASE_OFFSET;
+  rx_fifo_bounds |= (rx_fifo_end & SPI_DEVICE_RXF_ADDR_LIMIT_MASK)
+                    << SPI_DEVICE_RXF_ADDR_LIMIT_OFFSET;
+
+  uint32_t tx_fifo_bounds = 0;
+  tx_fifo_bounds |= (tx_fifo_start & SPI_DEVICE_TXF_ADDR_BASE_MASK)
+                    << SPI_DEVICE_TXF_ADDR_BASE_OFFSET;
+  tx_fifo_bounds |= (tx_fifo_end & SPI_DEVICE_TXF_ADDR_LIMIT_MASK)
+                    << SPI_DEVICE_TXF_ADDR_LIMIT_OFFSET;
+
+  spi->rx_fifo_base = rx_fifo_start;
+  spi->rx_fifo_len = config->rx_fifo_len;
+  spi->tx_fifo_base = tx_fifo_start;
+  spi->tx_fifo_len = config->tx_fifo_len;
+
+  // Now that we know that all function arguments are valid, we can abort
+  // all current transactions and begin reconfiguring the device.
+  dif_spi_device_result_t err;
+  err = dif_spi_device_abort(spi);
+  if (err != kDifSpiDeviceResultOk) {
+    return err;
+  }
+  err = dif_spi_device_irq_reset(spi);
+  if (err != kDifSpiDeviceResultOk) {
+    return err;
+  }
+
+  mmio_region_write32(spi->base_addr, SPI_DEVICE_CFG_REG_OFFSET, device_config);
+  mmio_region_write32(spi->base_addr, SPI_DEVICE_RXF_ADDR_REG_OFFSET,
+                      rx_fifo_bounds);
+  mmio_region_write32(spi->base_addr, SPI_DEVICE_TXF_ADDR_REG_OFFSET,
+                      tx_fifo_bounds);
+
+  return kDifSpiDeviceResultOk;
+}
+
+dif_spi_device_result_t dif_spi_device_abort(const dif_spi_device_t *spi) {
+  if (spi == NULL) {
+    return kDifSpiDeviceResultBadArg;
+  }
+
+  // Set the `abort` bit, and then spin until `abort_done` is asserted.
+  mmio_region_nonatomic_set_bit32(spi->base_addr, SPI_DEVICE_CONTROL_REG_OFFSET,
+                                  SPI_DEVICE_CONTROL_ABORT);
+  while (!mmio_region_get_bit32(spi->base_addr, SPI_DEVICE_STATUS_REG_OFFSET,
+                                SPI_DEVICE_STATUS_ABORT_DONE)) {
+  }
+
+  return kDifSpiDeviceResultOk;
+}
+
+dif_spi_device_result_t dif_spi_device_irq_reset(const dif_spi_device_t *spi) {
+  if (spi == NULL) {
+    return kDifSpiDeviceResultBadArg;
+  }
+
+  dif_spi_device_result_t err;
+
+  err = dif_spi_device_irq_clear_all(spi);
+  if (err != kDifSpiDeviceResultOk) {
+    return err;
+  }
+  err = dif_spi_device_set_irq_levels(spi, 0x80, 0x00);
+  if (err != kDifSpiDeviceResultOk) {
+    return err;
+  }
+
+  // Disable interrupts manually, since the relevant DIF only allows for single
+  // bit flips.
+  mmio_region_write32(spi->base_addr, SPI_DEVICE_INTR_ENABLE_REG_OFFSET, 0);
+
+  return kDifSpiDeviceResultOk;
+}
+
+dif_spi_device_result_t dif_spi_device_irq_get(
```

Comment:
The comment of this function says "Checks whether a particular interrupt has been enabled or disabled." but the implementation below returns if the device has a pending interrupt. The comments in the header file should be updated and we should have a separate function for getting the enabledness of IRQs.

Reply:
I already fixed this in some other version; I got myself mixed up on what this register did at some point.